File: works/1753.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
import java.util.PriorityQueue;

class Pair implements Comparable<Pair>
{
    int distance, cur;
    public Pair(int _dis, int _cur) { distance = _dis; cur = _cur; }

    @Override
    public int compareTo(Pair target) {
        return this.distance < target.distance ? 1 : -1;
    }
}

public class Main
{
    static int V, E, start;
    static int INF = 2000000000;

    static void Dijkstra(ArrayList<Integer> d, ArrayList<Pair>[] graph)
    {
        d.set(start, 0);
        PriorityQueue<Pair> pq = new PriorityQueue<>();
        pq.add(new Pair(0, start));

        while(!pq.isEmpty())
        {
            Pair top = pq.poll();
            int distance = -top.distance;
            int cur = top.cur;

            if(d.get(cur) < distance) continue;

            for(int i = 0; i < graph[cur].size(); i++)
            {
                int next = graph[cur].get(i).distance;
                int nextDistance = distance + graph[cur].get(i).cur;

                if(nextDistance < d.get(next))
                {
                    d.set(next, nextDistance);
                    pq.add(new Pair(-nextDistance, next));
                }
            }
        }


    }

    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] _s = br.readLine().split(" ");
        V = Integer.parseInt(_s[0]);
        E = Integer.parseInt(_s[1]);
        start = Integer.parseInt(br.readLine());

        ArrayList<Pair>[] graph = new ArrayList[V+1];
        for(int i = 0; i < V+1; i++) graph[i] = new ArrayList<>();
        for(int i = 0; i < E; i++)
        {
            _s = br.readLine().split(" ");
            graph[Integer.parseInt(_s[0])].add(new Pair(Integer.parseInt(_s[1]), Integer.parseInt(_s[2])));
        }

        ArrayList<Integer> d = new ArrayList<>();
        for(int i = 0; i < V+1; i++) d.add(INF);
        Dijkstra(d, graph);

        for(int i = 1; i <= V; i++)
        {
            if(d.get(i) == INF) System.out.println("INF");
            else System.out.println(d.get(i));
        }
    }


}

File: works/1753.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class Pair(_distance:Int, _cur:Int) : Comparable<Pair>
{
    var distance = _distance
    var cur = _cur
    override fun compareTo(other: Pair): Int = other.distance - distance

}

var V = 0; var E = 0; var start = 0
val INF = 2000000000

fun Dijkstra(d: ArrayList<Int>, graph: Array<ArrayList<Pair>>)
{
    d[start] = 0
    var pq = PriorityQueue<Pair>()
    pq.add(Pair(0, start))

    while(!pq.isEmpty())
    {
        var top = pq.poll()
        var distance = -top.distance
        var cur = top.cur

        if(d[cur] < distance) continue

        for(i in 0 until graph[cur].size)
        {
            var next = graph[cur][i].distance
            var nextDistance = distance + graph[cur][i].cur

            if(nextDistance < d[next])
            {
                d[next] = nextDistance
                pq.add(Pair(-nextDistance, next))
            }
        }
    }
}

fun main()
{
    val br = BufferedReader(InputStreamReader(System.`in`))
    var _s = br.readLine().split(' ')
    V = _s[0].toInt()
    E = _s[1].toInt()
    start = br.readLine().toInt()

    var graph = Array(V+1) { ArrayList<Pair>()}
    for(i in 0 until E)
    {
        _s = br.readLine().split(' ')
        graph[_s[0].toInt()].add(Pair(_s[1].toInt(), _s[2].toInt()))
    }

    var d = ArrayList<Int>()
    for(i in 0 until V+1) d.add(INF)
    Dijkstra(d, graph)

    var ans = ""
    for(i in 1 .. V)
    {
        if(d[i] == INF) ans += "INF\n"
        else { ans += d[i]; ans += '\n' }
    }

    print(ans)
}